In [53]:
# DATOS
from ucimlrepo import fetch_ucirepo 
# TRATAMIENTO DE DATOS
import pandas as pd 
import numpy as np
from utils import explore_data
from utils import explore_data2
# from utils import data_report
import math

# VISUALIZACIÓN
import matplotlib.pyplot as plt

# data (as pandas dataframes) 
df = pd.read_csv('./data/auto-mpg.csv')
  

## PRIMERA EXPLORACIÓN

In [9]:
df.head()

mpg  cylinders  displacement horsepower  weight  acceleration  model year  \
0  18.0          8         307.0        130    3504          12.0          70   
1  15.0          8         350.0        165    3693          11.5          70   
2  18.0          8         318.0        150    3436          11.0          70   
3  16.0          8         304.0        150    3433          12.0          70   
4  17.0          8         302.0        140    3449          10.5          70   

   origin                   car name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

In [39]:
explore_data(df)

----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Data Set Shape: (398, 9)
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

1. El DF MPG tiene 9 columnas y 397 filas de las cuales

    -**mpg**: significa millas por galón.
    - cylinders: cantidad de cilindros que tiene el coche.
    - displacement: centímetros cúbicos del coche.
    - horsepower: caballos de fuerza.
    - weight: peso del coche.
    - acceleration: aceleración.
    - model year: año de creación.
    - origin: continente de origen.
    - car name: nombre del modelo de coche.

2. Todas las variables son ``numéricas`` excepto *horsepower* y *car name*
3. No hay ``nulos``

In [44]:
explore_data2(df)

Primeras filas del DataFrame:
    mpg  cylinders  displacement horsepower  weight  acceleration  model year  \
0  18.0          8         307.0        130    3504          12.0          70   
1  15.0          8         350.0        165    3693          11.5          70   
2  18.0          8         318.0        150    3436          11.0          70   
3  16.0          8         304.0        150    3433          12.0          70   
4  17.0          8         302.0        140    3449          10.5          70   

   origin                   car name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino  

Información del DataFrame:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null

In [18]:
print(df.duplicated().sum())
print(df.duplicated().value_counts())
df[df.duplicated()].sort_values(by='car name')
df.loc[df.duplicated()]

0
False    398
Name: count, dtype: int64


Empty DataFrame
Columns: [mpg, cylinders, displacement, horsepower, weight, acceleration, model year, origin, car name]
Index: []

In [36]:
df.sample(10)

mpg  cylinders  displacement horsepower  weight  acceleration  \
91   13.0          8         400.0        150    4464          12.0   
207  20.0          4         130.0        102    3150          15.7   
347  37.0          4          85.0         65    1975          19.4   
54   35.0          4          72.0         69    1613          18.0   
23   26.0          4         121.0        113    2234          12.5   
191  22.0          6         225.0        100    3233          15.4   
11   14.0          8         340.0        160    3609           8.0   
354  34.5          4         100.0          ?    2320          15.8   
161  16.0          6         250.0        105    3897          18.5   
376  37.0          4          91.0         68    2025          18.2   

     model year  origin                   car name  
91           73       1  chevrolet caprice classic  
207          76       2                  volvo 245  
347          81       3             datsun 210 mpg  
54           71       3                datsun 1200  
23           70       2                   bmw 2002  
191          76       1           plymouth valiant  
11           70       1         plymouth 'cuda 340  
354          81       2                renault 18i  
161          75       1  chevroelt chevelle malibu  
376          82       3         mazda glc custom l

In [38]:
df.describe(include='all')

mpg   cylinders  displacement horsepower       weight  \
count   398.000000  398.000000    398.000000        398   398.000000   
unique         NaN         NaN           NaN         94          NaN   
top            NaN         NaN           NaN        150          NaN   
freq           NaN         NaN           NaN         22          NaN   
mean     23.514573    5.454774    193.425879        NaN  2970.424623   
std       7.815984    1.701004    104.269838        NaN   846.841774   
min       9.000000    3.000000     68.000000        NaN  1613.000000   
25%      17.500000    4.000000    104.250000        NaN  2223.750000   
50%      23.000000    4.000000    148.500000        NaN  2803.500000   
75%      29.000000    8.000000    262.000000        NaN  3608.000000   
max      46.600000    8.000000    455.000000        NaN  5140.000000   

        acceleration  model year      origin    car name  
count     398.000000  398.000000  398.000000         398  
unique           NaN         NaN         NaN         305  
top              NaN         NaN         NaN  ford pinto  
freq             NaN         NaN         NaN           6  
mean       15.568090   76.010050    1.572864         NaN  
std         2.757689    3.697627    0.802055         NaN  
min         8.000000   70.000000    1.000000         NaN  
25%        13.825000   73.000000    1.000000         NaN  
50%        15.500000   76.000000    1.000000         NaN  
75%        17.175000   79.000000    2.000000         NaN  
max        24.800000   82.000000    3.000000         NaN

1. **Horse power** tiene 94 únicos y el valor ``150`` es el que más se repite.
2. **Car name** tiene 305 únicos y el valos que más se repite es ``ford pinto``.
3. **columnas numéricas** no siguen una aparente distribución normal.

In [47]:
df.dtypes

mpg             float64
cylinders         int64
displacement    float64
horsepower       object
weight            int64
acceleration    float64
model year        int64
origin            int64
car name         object
dtype: object

|Variable| tipo | método| 
|---|----|-----|
|mpg| continua| mirar si todos terminan en 0, si es así, pasar a int |
|cylinders | discreta| ver los únicos y un conteo|
|displacement| continua| mirar si todos terminan en 0, si es así, pasar a int|
|horsepower   | discreta|convertir en int|
|weight| continua(discreta con muchas categorías)| nada |
|acceleration| continua|nada  |
|model year | discreta| agrupar por decada|
|origin|discreta|nada  |
|car name |categórica nominal|buscar la marca y elminar el resto |

In [64]:
   # Sacamos los NOMBRES
cols = pd.DataFrame(df.columns.values, columns=["COL_N"])

    # Sacamos los TIPOS
types = pd.DataFrame(df.dtypes.values, columns=["DATA_TYPE"])

    # Sacamos los MISSINGS
percent_missing = round(df.isnull().sum() * 100 / len(df), 2)
percent_missing_df = pd.DataFrame(percent_missing.values, columns=["MISSINGS (%)"])

    # Sacamos los VALORES UNICOS
unicos = pd.DataFrame(df.nunique().values, columns=["UNIQUE_VALUES"])
    
percent_cardin = round(unicos['UNIQUE_VALUES']*100/len(df), 2)
percent_cardin_df = pd.DataFrame(percent_cardin.values, columns=["CARDIN (%)"])

concatenado = pd.concat([cols, types, percent_missing_df, unicos, percent_cardin_df], axis=1, sort=False)
concatenado.set_index('COL_N', drop=True, inplace=True)

In [65]:
concatenado.T

COL_N            mpg cylinders displacement horsepower weight acceleration  \
DATA_TYPE      int32     int64      float64     object  int64      float64   
MISSINGS (%)     0.0       0.0          0.0        0.0    0.0          0.0   
UNIQUE_VALUES     36         5           82         94    351           95   
CARDIN (%)      9.05      1.26         20.6      23.62  88.19        23.87   

COL_N         model year origin car name  
DATA_TYPE          int64  int64   object  
MISSINGS (%)         0.0    0.0      0.0  
UNIQUE_VALUES         13      3      305  
CARDIN (%)          3.27   0.75    76.63

## LIMPIEZA


In [59]:
df[~df['mpg'].astype(str).str.endswith('.0')]


df['mpg'] = df['mpg'].astype(int)# si intento cambiar de float a int y me deja, es que no había decimales (no debería) EN ESTE CASO SI LO ESTÁ HACIENDO

In [56]:
df['mpg'].apply(lambda x: x - math.trunc(x))# restamos la parte entera a todos los datos


0      0.0
1      0.0
2      0.0
3      0.0
4      0.0
      ... 
393    0.0
394    0.0
395    0.0
396    0.0
397    0.0
Name: mpg, Length: 398, dtype: float64

In [60]:
df[~df['displacement'].astype(str).str.endswith('.0')]

mpg  cylinders  displacement horsepower  weight  acceleration  model year  \
58   25          4          97.5         80    2126          17.0          72   

    origin            car name  
58       1  dodge colt hardtop

In [ ]:
df.drop(columns=['car name'], inplace=True)

In [66]:
df['horsepower'] = df['horsepower'].astype(int)

ValueError: invalid literal for int() with base 10: '?'

In [68]:
# pd.set_option('display.max_rows', None)
# df.horsepower.value_counts().reset_index()
# forma fea

horsepower  count
0         150     22
1          90     20
2          88     19
3         110     18
4         100     17
5          75     14
6          95     14
7         105     12
8          70     12
9          67     12
10         65     10
11         97      9
12         85      9
13         80      7
14        145      7
15        140      7
16         92      6
17         68      6
18         78      6
19         84      6
20          ?      6
21         72      6
22        180      5
23         71      5
24         86      5
25        115      5
26        130      5
27         60      5
28        170      5
29        175      5
30         83      4
31        120      4
32         76      4
33        165      4
34         52      4
35        190      3
36        125      3
37        215      3
38         48      3
39         63      3
40        112      3
41        225      3
42         96      3
43         69      3
44         74      3
45        129      2
46         98      2
47        198      2
48         53      2
49         81      2
50         79      2
51        153      2
52         58      2
53         62      2
54         87      2
55         46      2
56        160      2
57        155      2
58        139      2
59         66      1
60        149      1
61        132      1
62         64      1
63        116      1
64         77      1
65         89      1
66        142      1
67        135      1
68        138      1
69        133      1
70        103      1
71        107      1
72        108      1
73        102      1
74        220      1
75        113      1
76        200      1
77        210      1
78        193      1
79         54      1
80        208      1
81        137      1
82        158      1
83        167      1
84         94      1
85        230      1
86         49      1
87         91      1
88        122      1
89         61      1
90         93      1
91        148      1
92        152      1
93         82      1

In [71]:
# for i in df['horsepower'] :
#     if i .isnumeric() == False:
#         print(i)

df['horsepower'] = pd.to_numeric(df['horsepower'], errors='coerce')  # convierte a NaN los valores que no son numericos. Tmabien está la opción de raise error.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    int32  
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        398 non-null    int64  
 5   acceleration  398 non-null    float64
 6   model year    398 non-null    int64  
 7   origin        398 non-null    int64  
 8   car name      398 non-null    object 
dtypes: float64(3), int32(1), int64(4), object(1)
memory usage: 26.6+ KB


In [ ]:
# df.horsepower.fillna(df.horsepower.median(), inplace=True)
# no sé ahora cómo lo voy a imputar, si con la media, mediana, si agrupo por marca etc. 

In [72]:
df.sample(20)
# para comprobar si la primera palabra de car name siempre es la marca

mpg  cylinders  displacement  horsepower  weight  acceleration  \
329   44          4          91.0        67.0    1850          13.8   
375   36          4         105.0        74.0    1980          15.3   
300   23          8         260.0        90.0    3420          22.2   
175   29          4          90.0        70.0    1937          14.0   
323   27          4         156.0       105.0    2800          14.4   
149   24          4         120.0        97.0    2489          15.0   
256   20          6         225.0       100.0    3430          17.2   
54    35          4          72.0        69.0    1613          18.0   
258   20          6         231.0       105.0    3380          15.8   
190   14          8         351.0       152.0    4215          12.8   
35    17          6         250.0       100.0    3329          15.5   
125   20          6         198.0        95.0    3102          16.5   
21    24          4         107.0        90.0    2430          14.5   
326   43          4          90.0        48.0    2335          23.7   
310   38          4          89.0        60.0    1968          18.8   
4     17          8         302.0       140.0    3449          10.5   
311   32          4          98.0        70.0    2120          15.5   
364   26          8         350.0       105.0    3725          19.0   
99    18          6         232.0       100.0    2945          16.0   
111   18          3          70.0        90.0    2124          13.5   

     model year  origin                           car name  
329          80       3                honda civic 1500 gl  
375          82       2                volkswagen rabbit l  
300          79       1  oldsmobile cutlass salon brougham  
175          75       2                  volkswagen rabbit  
323          80       1                         dodge colt  
149          74       3                        honda civic  
256          78       1                    plymouth volare  
54           71       3                        datsun 1200  
258          78       1              buick century special  
190          76       1                   ford gran torino  
35           71       1          chevrolet chevelle malibu  
125          74       1                    plymouth duster  
21           70       2                        audi 100 ls  
326          80       2                 vw dasher (diesel)  
310          80       3              toyota corolla tercel  
4            70       1                        ford torino  
311          80       1                 chevrolet chevette  
364          81       1              oldsmobile cutlass ls  
99           73       1                         amc hornet  
111          73       3                          maxda rx3

In [73]:
# buscamos la marca 
df['marca']= df['car name'].apply(lambda x: x.split()[0])
df['marca'].value_counts()

marca
ford             51
chevrolet        43
plymouth         31
amc              28
dodge            28
toyota           25
datsun           23
buick            17
pontiac          16
volkswagen       15
honda            13
mercury          11
mazda            10
oldsmobile       10
fiat              8
peugeot           8
audi              7
chrysler          6
vw                6
volvo             6
renault           5
saab              4
subaru            4
opel              4
chevy             3
bmw               2
cadillac          2
maxda             2
mercedes-benz     2
triumph           1
vokswagen         1
mercedes          1
hi                1
capri             1
chevroelt         1
toyouta           1
nissan            1
Name: count, dtype: int64

In [74]:
df['marca'].replace({
    'chevroelt':'chevrolet',
    'chevy':'chevrolet',
    'capri':'ford',
    'maxda':'mazda', 
    'toyouta':'toyota', 
    'vw':'volkswagen',
    'vokswagen':'volkswagen',
    'mercedes':'mercedes-benz'}, inplace=True)
df['marca'].value_counts().reset_index()

marca  count
0            ford     52
1       chevrolet     47
2        plymouth     31
3             amc     28
4           dodge     28
5          toyota     26
6          datsun     23
7      volkswagen     22
8           buick     17
9         pontiac     16
10          honda     13
11          mazda     12
12        mercury     11
13     oldsmobile     10
14        peugeot      8
15           fiat      8
16           audi      7
17       chrysler      6
18          volvo      6
19        renault      5
20         subaru      4
21           opel      4
22           saab      4
23  mercedes-benz      3
24            bmw      2
25       cadillac      2
26        triumph      1
27             hi      1
28         nissan      1

In [76]:
df.drop(columns=['car name'], inplace=True)

## EDA

In [77]:
for col in df.columns:
    if df[col].dtype != 'O' and col !='weight':
        analisis_univariante_con_tabla_completa(df, col);  # funcion de juan que hace un analisis de cada columna de forma univariante. Histograma, boxplot y sample quartiles.

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0     18          8         307.0       130.0    3504          12.0   
1     15          8         350.0       165.0    3693          11.5   
2     18          8         318.0       150.0    3436          11.0   
3     16          8         304.0       150.0    3433          12.0   
4     17          8         302.0       140.0    3449          10.5   
5     15          8         429.0       198.0    4341          10.0   
6     14          8         454.0       220.0    4354           9.0   
7     14          8         440.0       215.0    4312           8.5   
8     14          8         455.0       225.0    4425          10.0   
9     15          8         390.0       190.0    3850           8.5   
10    15          8         383.0       170.0    3563          10.0   
11    14          8         340.0       160.0    3609           8.0   
12    15          8         400.0       150.0    3761           9.5   
13    14          8         455.0       225.0    3086          10.0   
14    24          4         113.0        95.0    2372          15.0   
15    22          6         198.0        95.0    2833          15.5   
16    18          6         199.0        97.0    2774          15.5   
17    21          6         200.0        85.0    2587          16.0   
18    27          4          97.0        88.0    2130          14.5   
19    26          4          97.0        46.0    1835          20.5   
20    25          4         110.0        87.0    2672          17.5   
21    24          4         107.0        90.0    2430          14.5   
22    25          4         104.0        95.0    2375          17.5   
23    26          4         121.0       113.0    2234          12.5   
24    21          6         199.0        90.0    2648          15.0   
25    10          8         360.0       215.0    4615          14.0   
26    10          8         307.0       200.0    4376          15.0   
27    11          8         318.0       210.0    4382          13.5   
28     9          8         304.0       193.0    4732          18.5   
29    27          4          97.0        88.0    2130          14.5   
30    28          4         140.0        90.0    2264          15.5   
31    25          4         113.0        95.0    2228          14.0   
32    25          4          98.0         NaN    2046          19.0   
33    19          6         232.0       100.0    2634          13.0   
34    16          6         225.0       105.0    3439          15.5   
35    17          6         250.0       100.0    3329          15.5   
36    19          6         250.0        88.0    3302          15.5   
37    18          6         232.0       100.0    3288          15.5   
38    14          8         350.0       165.0    4209          12.0   
39    14          8         400.0       175.0    4464          11.5   
40    14          8         351.0       153.0    4154          13.5   
41    14          8         318.0       150.0    4096          13.0   
42    12          8         383.0       180.0    4955          11.5   
43    13          8         400.0       170.0    4746          12.0   
44    13          8         400.0       175.0    5140          12.0   
45    18          6         258.0       110.0    2962          13.5   
46    22          4         140.0        72.0    2408          19.0   
47    19          6         250.0       100.0    3282          15.0   
48    18          6         250.0        88.0    3139          14.5   
49    23          4         122.0        86.0    2220          14.0   
50    28          4         116.0        90.0    2123          14.0   
51    30          4          79.0        70.0    2074          19.5   
52    30          4          88.0        76.0    2065          14.5   
53    31          4          71.0        65.0    1773          19.0   
54    35          4          72.0        69.0    1613          18.0   
55    27          4          

Hay que escribir en cada columna una pequeña descripción de su 
continua: se dice el rango
discreta: se dice el max y minimo.
mediana sale .5si es par o numero entero si es impar

si media = moda --- normal
valores cercanos a o en skew y kurtosis es q está muy centrada.
distribuciónn para variables discretas no tiene mucho sentido. si los datos están valanceados la linea de distrib deberia ser recta


max y minimo de los boxplot : lineas rango intercuartilico *1.5 se suma hacia arriba y se resta hacia abajo.
se cortan los outliers a partir del maximo del boxplot